Fixed useQuery under StrictMode creating 2 observable query

Cellule opened the pull request:
Closes #11914

<!--
  Thanks for filing a pull request on Apollo Client!

  A few automated bots may chime in on your PR. They are here to help
  with reviewing and ensuring Apollo Client is production ready after each
  pull request merge.

    - apollo-cla will respond asking you to sign the CLA if this is your first PR.
      It may also respond with warnings, messages, or fail the build if something is off.
      Don't worry, it'll help you to fix what is broken!

    - bundlesize is a status check to keep the footprint of Apollo Client as small as possible.

    - circleci will run tests, checking style of code, and generally make
      sure everything is working as expected

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:

### Checklist:
- [X] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests
-->

jerelmiller reviewed (changes requested):
I believe the fix you presented here is a reasonable one and I'm good with that fix!

Where I'd like to see a change potentially is how we setup the test to reproduce this behavior. See my comments on the tests and let me know what you think!

Cellule commented on src/core/__tests__/QueryManager/index.ts:
```diff
@@ -4878,6 +4878,44 @@ describe("QueryManager", () => {
+        });
+        // Recreate state where an observable query is dirty but has no observers in the query manager
+        // @ts-expect-error -- Accessing private field for testing
+        oq.queryInfo.dirty = true;
```
My idea was to reproduce the bad state, regardless of what causes the bad state
I agree that it's better to do a fully functional test especially to identify the "real" cause of the issue, but I haven't had time to look into it this week because I was swamped and I'm going on vacations starting next week.

If you want to pick up the PR to wrap it up I don't mind. 
Otherwise I'll try to look into it when I get a chance (it's still on my list 😅)

Cellule replied:
@jerelmiller it was great to chat with you yesterday at the GraphQL Summit!
I rebased my fix and gave a shot at writing a test with useQuery.
I have not been able to reproduce the issue in the test however :(
Happy to get your input on my test how you think we can bring this to the finish line

svc-apollo-docs commented:
### ❌ Docs Preview Failed

Error
```
HttpError: Resource not accessible by personal access token - https://docs.github.com/rest/commits/statuses#create-a-commit-status
```

Cellule commented on src/core/QueryManager.ts:
```diff
@@ -880,10 +880,7 @@ export class QueryManager<TStore> {
 
-        if (
-          fetchPolicy === "standby" ||
-          (include === "active" && !oq.hasObservers())
```
This part is related to this part of the issue my colleague found

https://github.com/apollographql/apollo-client/issues/11914#issuecomment-2407750879

This code is 3 years old, so I'm skeptical that my change makes sense.
From my perspective, if someone provided a list of DocumentNode or OperationName to refetch, they clearly meant to refetch only `active` ones.
I can't think of a reason why you'd want to refetch queries without any observers

Cellule replied:
Humm looks like I was wrong
https://github.com/Cellule/apollo-client/blob/1e23e8e65029f0d2eb7b34e1f48b24f7329fbd00/src/__tests__/refetchQueries.ts#L535-L538
I still don't know why you'd want that, but there must be a good reason.
This comes back to why we end up with queries without observers with `useQuery`

jerelmiller replied:
Huh, I didn't realize that was a thing. Let me ask around and see if I can figure out why we'd ever want this. I agree that fetching a query with no observers is an odd behavior 🤔

jerelmiller replied:
I'm noodling on the rest of the implementation, but I'm curious... with those other changes, do we need to modify this? If an `ObservableQuery` is created but not registered, it shouldn't be included in `this.queries` correct?

I'm wondering if we should leave this alone for now. To be honest, I'm not sure how much of a breaking change this is and while I think it makes sense to you and I that this should be the correct behavior, we've had fiery discussions on far smaller changes 😂. We will be starting work on v4 here very soon, so I think we could probably punt this change to there where we can announce this change more formally.

Cellule replied:
Fair, I don't mind living with this part as a patch on our side and maybe we can surface up reasons to not do it XD
If we manage to completely solve the StrictMode issue so that QueryManager never has "dead" queries then this change is not as important anymore either

jerelmiller replied:
Yep exactly!

Cellule commented:
@jerelmiller Let me know how you feel about the direction I've taken in my PR to fix the StrictMode issue before I dig in more into the current test failures

jerelmiller commented on src/core/__tests__/QueryManager/index.ts:
```diff
@@ -4712,6 +4712,46 @@ describe("QueryManager", () => {
+        });
+        // Recreate state where an observable query is dirty but has no observers in the query manager
+        // @ts-expect-error -- Accessing private field for testing
+        oq.queryInfo.dirty = true;
```
I'm curious, now that we've sorta pinned this down on both strict mode and the creation of multiple `ObservableQuery` instances without observers, do we have a need for this condition anymore?

jerelmiller reviewed (changes requested):
Took me a bit to come around to this solution, but I kind of dig it. I'd like to discuss this with the team in our next team meeting as well in case someone else has a suggestion that I'm not thinking of, but otherwise I think this is a positive change.

For the changes there now, I do have a few suggestions that I think could simplify a touch more without the need to pass around the `queryInfo` or `register` function. Let me know what you think!

jerelmiller commented:
I asked the team about this and my colleague pointed out something that we do in `useSubscription` that we may or may not be able to do here. Check out this bit of code:

https://github.com/apollographql/apollo-client/blob/6496b25613a10c7d99633d1ceef8c895d224d527/src/react/hooks/useSubscription.ts#L354-L363

Here we wrap the subscription with its own `Observable` so that we only create the `ObservableQuery` when its subscribed. Not sure if you want to perhaps noodle on this? This might bring its own challenges with `useQuery`, but if you're able to get something working like this then we can isolate the changes to `useQuery` itself without having to touch the core API.

Thoughts?

jerelmiller commented on src/core/QueryInfo.ts:
```diff
@@ -287,11 +287,17 @@ export class QueryInfo {
+      !this.listeners.size ||
+      // It's possible that the query is no longer being watched, but the
+      // ObservableQuery is still active/pending cleanup. In this case, we should not notify.
+      !this.observableQuery?.hasObservers()
```
After talking with the team, I think we might want to move this change to 4.0 and make it a breaking change so that we can call it out prominently. Its difficult to tell who might rely on this behavior if you're using the core API directly. I think the fix for strict mode might be enough here to address what you're seeing specifically in your app. Moving this to a major allows us to ensure we add this behavior everywhere as well.

As an FYI, we will begin work on 4.0 once 3.12 ships (a few weeks from now), so shouldn't be long before we can address this 🙂

Cellule commented:
> I asked the team about this and my colleague pointed out something that we do in `useSubscription` that we may or may not be able to do here. Check out this bit of code:
> 
> https://github.com/apollographql/apollo-client/blob/6496b25613a10c7d99633d1ceef8c895d224d527/src/react/hooks/useSubscription.ts#L354-L363
> 
> Here we wrap the subscription with its own `Observable` so that we only create the `ObservableQuery` when its subscribed. Not sure if you want to perhaps noodle on this? This might bring its own challenges with `useQuery`, but if you're able to get something working like this then we can isolate the changes to `useQuery` itself without having to touch the core API.
> 
> Thoughts?

So I was digging into this, I haven't able to easily replicate that kind of behavior for useQuery.
My biggest issue was that too much code relies on the `ObservableQuery` to be readily available and returned to the user through `obsQueryFields`
It would likely be possible to make it work, but as I started hacking at it, it got messy fast.

So instead I considered another angle, since we want to register the query in the QueryManager when the ObservableQuery subscribes, why not let it take care of subscribing.
The ObservableQuery is also in charge of removing itself through `tearDownQuery` so I thought it made sense.
I removed `this.queries.set(observable.queryId, queryInfo);` from `QueryManager.watchQuery` and added `this.queryManager["queries"].set(this.queryId, this.queryInfo);` in the subscribe function of ObservableQuery.
It worked in my scenario, but I noticed that a few other tests failed, namely in `createQueryPreloader` and some others. The tests looked fine, so I was suspicious.
I'm not sure why you would want to register a query that you won't subscribe to, but anyway.

So now I'm looking at a solution to have this behavior only in useQuery.
I passed a hidden option through WatchQueryOption that would only set the query in QueryManager.queries in the ObservableQuery subscribe function.
It seems to work, but look a bit hackish (maybe less than my current solution?)

Boy oh boy, I thought the solution worked because I only tested on React19, but it seems on React18 and 17 the behavior of `useSyncExternalStore` causes to call the `getSnapshot` function breaking React rules which registers the query in the ApolloCache
When another query gets result, it would broadcast to all cache watches causing the query to reobserve.
I initially hid/mitigated this part with the `shouldNotify` "fix"

Now I really feel like we might need to go down the path of only creating the `ObservableQuery` when it actually subscribes 😮‍💨

jerelmiller commented:
Gah. I had another conference to attend then was on vacation. Apologies for my slow time to respond!

I will take a look at your changes and respond to your latest comment next week. Thanks so much for your patience!

Cellule commented:
Awesome! I'll definitely give it a try in the near future :D
